lib/vrl/stdlib/src/insert.rs

```diff
@@ -0,0 +1,140 @@
+            Example {
+                title: "nested fields",
+                source: r#"insert!({}, ["foo", "bar"], "baz")"#,
+                result: Ok(r#"{ "foo": { "bar" : "baz" } }"#),
+            },
+            Example {
+                title: "indexing",
+                source: r#"insert!({"foo": { "bar": [] }}, ["foo", "bar", 1], "baz")"#,
+                result: Ok(r#"{ "foo": { "bar": [null, "baz"] } }"#),
+            },
+            Example {
+                title: "external target",
+                source: indoc! {r#"
+                    . = { "foo": true }
+                    insert!(., "bar", "baz")
+                "#},
+                result: Ok(r#"{ "foo": true, "bar": "baz" }"#),
+            },
+            Example {
+                title: "variable",
+                source: indoc! {r#"
+                    var = { "foo": true }
+                    insert!(var, "bar", "baz")
+                "#},
+                result: Ok(r#"{ "foo": true, "bar": "baz" }"#),
+            },
+            Example {
+                title: "invalid segment type",
+                source: r#"insert!({"foo": { "bar": [92, 42] }}, ["foo", true], "baz")"#,
+                result: Err(
+                    r#"function call error for "insert" at (0:59): expected "string" or "integer", got "boolean""#,
+                ),
+            },
+        ]
+    }
+
+    fn compile(&self, _state: &state::Compiler, mut arguments: ArgumentList) -> Compiled {
+        let target = arguments.required("target");
+        let field = arguments.required("field");
+        let value = arguments.required("value");
+
+        Ok(Box::new(InsertFn {
+            target,
+            field,
+            value,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct InsertFn {
+    target: Box<dyn Expression>,
+    field: Box<dyn Expression>,
+    value: Box<dyn Expression>,
+}
+
+impl Expression for InsertFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let path = match self.field.resolve(ctx)? {
+            Value::Bytes(path) => LookupBuf::from(String::from_utf8_lossy(&path)),
+            Value::Array(paths) => {
+                let mut lookup = LookupBuf::root();
+
+                for path in paths {
+                    let segment = match path {
+                        Value::Bytes(path) => {
+                            SegmentBuf::Field(String::from_utf8_lossy(&path).into_owned().into())
+                        }
+                        Value::Integer(index) => SegmentBuf::Index(index as isize),
+                        value => {
+                            return Err(value::Error::Expected {
+                                got: value.kind(),
+                                expected: Kind::Bytes | Kind::Integer,
+                            }
+                            .into())
+                        }
+                    };
+
+                    lookup.push_back(segment)
+                }
```

Comment:
This maybe seems like something a `Lookup` should know how to do (build itself from a list of parts).

Reply:
Maybe, yes. I wanted to avoid touching the existing `Lookup` too much, since I believe there's still something on the roadmap to make it interface better between VRL and Vector.

Even so, Lookup doesn't know about VRL's `Value` type, so you'd still have to do the looping and matching here, so I'm not sure how much there's still left to gain at that point.